### ISSUE TITLE
Epic: FinanceControll - Privacy-First Investment Tracker

### ISSUE THREAD

**Issue by vongohren** [vongohren/FinanceControll#1]
# FinanceControll - Privacy-First Investment Tracker

> **"Your data, your control. Local by default, cloud by choice."**

## Vision

A personal finance tracker for investments in startups, funds, state obligations, crypto, and other assets. Built with privacy at the core - data stays local unless you choose otherwise.

Inspired by [Ghostfolio](https://ghostfol.io) and [Wealthfolio](https://github.com/afadil/wealthfolio), but with a unique focus on **illiquid investments** (startup equity, fund commitments) alongside traditional assets.

---

## Core Philosophy

1. **Local-first** - Data stays on device unless you choose otherwise
2. **Zero-knowledge option** - Hosted service never sees unencrypted data
3. **Portable** - Export everything, no lock-in
4. **Open source** - Auditable, forkable, trustable

---

## Tech Stack

| Component | Technology | Rationale |
|-----------|-----------|-----------|
| Runtime | **Bun** | Fast, native TypeScript, modern tooling |
| Framework | **Next.js 15** (App Router) | Server Components, Server Actions, Vercel-ready |
| Local DB | **PGlite** | Postgres in WASM, same schema everywhere |
| Cloud DB | **Supabase** (PostgreSQL) | Managed Postgres, Auth, RLS |
| ORM | **Drizzle ORM** | Type-safe, SQL-native, lightweight |
| UI | **shadcn/ui + Tailwind** | Accessible, customizable |
| Charts | **Recharts** | Time series, pie charts, flexible |
| Currency API | **Exchange Rate API** | Live rates with historical capture |

---

## Architecture: Pluggable Storage Layer

```
┌─────────────────────────────────────────────────────────────────┐
│                     Application Layer                           │
│              (Next.js App Router + React UI)                    │
└─────────────────────────────────────────────────────────────────┘
                               │
┌─────────────────────────────────────────────────────────────────┐
│                    Repository Layer                              │
│   portfolioRepo | assetRepo | transactionRepo | valuationRepo   │
│            (Same interface regardless of storage)                │
└─────────────────────────────────────────────────────────────────┘
                               │
┌─────────────────────────────────────────────────────────────────┐
│                  Storage Adapter Interface                       │
│                                                                  │
│   ┌──────────────┐  ┌──────────────┐  ┌──────────────────────┐  │
│   │  PGlite      │  │  Postgres    │  │  Supabase            │  │
│   │  (Local)     │  │  (Self-Host) │  │  (Cloud/Hosted)      │  │
│   ├──────────────┤  ├──────────────┤  ├──────────────────────┤  │
│   │ Browser:     │  │ Docker       │  │ Auth + RLS           │  │
│   │  IndexedDB   │  │ compose      │  │ managed              │  │
│   │ Desktop:     │  │ or any       │  │                      │  │
│   │  File system │  │ Postgres URL │  │ (Your own or hosted) │  │
│   └──────────────┘  └──────────────┘  └──────────────────────┘  │
└─────────────────────────────────────────────────────────────────┘
```

---

## Storage Adapter Interface

```typescript
// lib/storage/types.ts
export interface StorageAdapter {
  connect(): Promise<void>;
  disconnect(): Promise<void>;
  isConnected(): boolean;
  query<T>(sql: string, params?: unknown[]): Promise<T[]>;
  db: DrizzleInstance;
  migrate(): Promise<void>;
  exportData(): Promise<ExportedData>;
  importData(data: ExportedData): Promise<void>;
}
```

---

## Database Schema

```sql
-- Asset types enum
asset_type: startup_equity | fund | state_obligation | crypto | public_equity | other

-- Portfolios
portfolios (
  id uuid PRIMARY KEY,
  user_id uuid,  -- NULL for local mode
  name text NOT NULL,
  description text,
  base_currency text DEFAULT 'NOK',
  created_at timestamptz,
  updated_at timestamptz
)

-- Assets
assets (
  id uuid PRIMARY KEY,
  portfolio_id uuid REFERENCES portfolios NOT NULL,
  type asset_type NOT NULL,
  name text NOT NULL,
  ticker text,
  notes text,
  created_at timestamptz,
  updated_at timestamptz
)

-- Transactions (cost basis)
transactions (
  id uuid PRIMARY KEY,
  asset_id uuid REFERENCES assets NOT NULL,
  type: 'buy' | 'sell',
  quantity decimal NOT NULL,
  price_per_unit decimal NOT NULL,
  currency text NOT NULL,
  exchange_rate decimal NOT NULL,  -- rate to base currency at time of tx
  date date NOT NULL,
  notes text,
  created_at timestamptz
)

-- Valuations (point-in-time updates)
valuations (
  id uuid PRIMARY KEY,
  asset_id uuid REFERENCES assets NOT NULL,
  value_per_unit decimal NOT NULL,
  currency text NOT NULL,
  exchange_rate decimal NOT NULL,
  valuation_date date NOT NULL,
  source text,  -- "409A", "funding round", "market", "manual"
  notes text,
  created_at timestamptz
)

-- Exchange rate cache
exchange_rates (
  id uuid PRIMARY KEY,
  from_currency text NOT NULL,
  to_currency text NOT NULL,
  rate decimal NOT NULL,
  date date NOT NULL,
  UNIQUE(from_currency, to_currency, date)
)
```

---

## Calculation Logic

```
For each asset:
  total_quantity = SUM(buys) - SUM(sells)
  cost_basis = SUM(quantity * price_per_unit * exchange_rate) for buys
  current_value = total_quantity * latest_valuation.value_per_unit * exchange_rate
  gain_loss = current_value - cost_basis
  gain_loss_percent = (gain_loss / cost_basis) * 100
```

---

## Deployment Modes

| Mode | Storage | Auth | Use Case |
|------|---------|------|----------|
| **Local Web** | PGlite + IndexedDB | None | Privacy purists |
| **Local Desktop** | PGlite + File | None | True offline |
| **Self-Hosted** | Docker + Postgres | Self-managed | Power users |
| **BYOD Cloud** | User's Supabase | Supabase Auth | Own data, our app |
| **Hosted Service** | Our Supabase | Our Auth | Convenience |

---

## Project Structure

```
finance-controll/
├── app/
│   ├── (onboarding)/setup/page.tsx
│   ├── (app)/
│   │   ├── dashboard/
│   │   ├── portfolios/
│   │   └── assets/
│   └── api/
├── lib/
│   ├── storage/
│   │   ├── types.ts
│   │   ├── pglite-adapter.ts
│   │   ├── postgres-adapter.ts
│   │   ├── supabase-adapter.ts
│   │   └── context.tsx
│   ├── db/schema.ts
│   ├── repositories/
│   └── export/
├── components/
├── docker/
└── docs/
```

---

### Phase 0: Foundation (Privacy Architecture)
- Task 0: Setup Bun environment
- Task 1: Create Next.js 15 project
- Task 2: Design StorageAdapter interface
- Task 3: Implement PGliteAdapter
- Task 4: Create Drizzle schema
- Task 5: Build onboarding flow

### Phase 1: Local-First Core
- Tasks 6-12: Repository layer, CRUD, navigation

### Phase 2: Calculations & Display
- Tasks 13-17: Cost basis, gain/loss, currency

### Phase 3: Data Portability
- Tasks 18-20: JSON/CSV export/import

### Phase 4: Dashboard
- Tasks 21-24: Charts and visualizations

### Phase 5: Additional Adapters
- Tasks 25-27: Postgres, Supabase adapters

### Phase 6: Deploy & Distribute
- Tasks 28-32: Docker, Vercel, docs, open source

---

## AI Context Instructions

When working on tasks from this epic:

1. **Read this epic first** for full context
2. **Check the task's parent phase** for ordering
3. **Use Bun** for all package management and running scripts
4. **Follow the storage adapter pattern** - never directly access DB
5. **PGlite first** - everything must work locally before cloud
6. **Drizzle Postgres dialect** - same schema for all adapters
7. **shadcn/ui components** - use `bunx shadcn@latest add`

---

## Privacy Guarantees

- Local mode: Zero network requests for data
- No telemetry or analytics
- Full JSON export for portability
- Open source and auditable
- Self-hostable with Docker
- BYOD Supabase support

---

## Related Issues

Child tasks will be linked below as they are created.
